internal/controllers/device_plugin_reconciler.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	return err
+}
+
+func (dprh *devicePluginReconcilerHelper) garbageCollect(ctx context.Context,
+	mod *kmmv1beta1.Module,
+	existingDS []appsv1.DaemonSet) error {
+
+	logger := log.FromContext(ctx)
+	deleted := make([]string, 0)
+	for _, ds := range existingDS {
+		if isOlderVersionUnusedDevicePluginDaemonset(&ds, mod.Spec.ModuleLoader.Container.Version) {
+			deleted = append(deleted, ds.Name)
+			if err := dprh.client.Delete(ctx, &ds); err != nil {
+				return fmt.Errorf("could not delete device plugin DaemonSet %s: %v", ds.Name, err)
+			}
+		}
+	}
+
+	logger.Info("garbage-collected device plugin's DaemonSets", "names", deleted)
+	return nil
+}
+
+func (dprh *devicePluginReconcilerHelper) handleModuleDeletion(ctx context.Context, existingDevicePluginDS []appsv1.DaemonSet) error {
+	// delete all the Device Plugin Daemonset, in order to allow worker pods to delete kernel modules
+	for _, ds := range existingDevicePluginDS {
+		err := dprh.client.Delete(ctx, &ds)
+		if err != nil {
+			return fmt.Errorf("failed to delete device-plugin Daemonset %s/%s: %v", ds.Namespace, ds.Name, err)
+		}
+	}
+	return nil
+}
+
+func (dprh *devicePluginReconcilerHelper) setKMMOMetrics(ctx context.Context) {
+	logger := log.FromContext(ctx)
+
+	mods := kmmv1beta1.ModuleList{}
+	err := dprh.client.List(ctx, &mods)
+	if err != nil {
+		logger.V(1).Info("failed to list KMMomodules for metrics", "error", err)
+		return
+	}
+
+	numModules := len(mods.Items)
+	numModulesWithBuild := 0
+	numModulesWithSign := 0
+	numModulesWithDevicePlugin := 0
+	for _, mod := range mods.Items {
+		if mod.Spec.DevicePlugin != nil {
+			numModulesWithDevicePlugin += 1
+		}
+		buildCapable, signCapable := isModuleBuildAndSignCapable(&mod)
+		if buildCapable {
+			numModulesWithBuild += 1
+		}
+		if signCapable {
+			numModulesWithSign += 1
+		}
+
+		if mod.Spec.ModuleLoader.Container.Modprobe.Args != nil {
+			modprobeArgs := strings.Join(mod.Spec.ModuleLoader.Container.Modprobe.Args.Load, ",")
+			dprh.metricsAPI.SetKMMModprobeArgs(mod.Name, mod.Namespace, modprobeArgs)
+		}
+		if mod.Spec.ModuleLoader.Container.Modprobe.RawArgs != nil {
+			modprobeRawArgs := strings.Join(mod.Spec.ModuleLoader.Container.Modprobe.RawArgs.Load, ",")
+			dprh.metricsAPI.SetKMMModprobeRawArgs(mod.Name, mod.Namespace, modprobeRawArgs)
+		}
+	}
+	dprh.metricsAPI.SetKMMModulesNum(numModules)
+	dprh.metricsAPI.SetKMMInClusterBuildNum(numModulesWithBuild)
+	dprh.metricsAPI.SetKMMInClusterSignNum(numModulesWithSign)
+	dprh.metricsAPI.SetKMMDevicePluginNum(numModulesWithDevicePlugin)
+}
+
+func (dprh *devicePluginReconcilerHelper) getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error) {
+	mod := kmmv1beta1.Module{}
+
+	if err := dprh.client.Get(ctx, namespacedName, &mod); err != nil {
+		return nil, fmt.Errorf("failed to get the kmmo module %s: %w", namespacedName, err)
+	}
+	return &mod, nil
+}
```

Comment:
Same questionable usefulness of this method IMHO.

Reply:
as mentioned above - i would leave for v2.1

Reply:
https://github.com/kubernetes-sigs/kernel-module-management/issues/603